File: engine/src/core/logging/Logger.cpp
#include "core/logging/Logger.h"

namespace subspace {

Logger& Logger::Instance()
{
    static Logger instance;
    return instance;
}

void Logger::SetMinimumLevel(LogLevel level)
{
    std::lock_guard<std::mutex> lock(_mutex);
    _minimumLevel = level;
}

LogLevel Logger::GetMinimumLevel() const
{
    std::lock_guard<std::mutex> lock(_mutex);
    return _minimumLevel;
}

void Logger::Log(LogLevel level, const std::string& category, const std::string& message)
{
    std::lock_guard<std::mutex> lock(_mutex);
    if (level < _minimumLevel) return;

    LogEntry entry;
    entry.timestamp = std::chrono::system_clock::now();
    entry.level = level;
    entry.category = category;
    entry.message = message;

    _recentLogs.push_back(entry);
    if (_recentLogs.size() > kMaxRecentLogs) {
        _recentLogs.erase(_recentLogs.begin());
    }

    // Console output
    const char* tag = "INFO";
    switch (level) {
        case LogLevel::Debug:    tag = "DEBUG"; break;
        case LogLevel::Info:     tag = "INFO";  break;
        case LogLevel::Warning:  tag = "WARN";  break;
        case LogLevel::Error:    tag = "ERROR"; break;
        case LogLevel::Critical: tag = "CRIT";  break;
    }
    std::cout << "[" << tag << "] [" << category << "] " << message << "\n";
}

void Logger::Debug(const std::string& category, const std::string& message)
{
    Log(LogLevel::Debug, category, message);
}

void Logger::Info(const std::string& category, const std::string& message)
{
    Log(LogLevel::Info, category, message);
}

void Logger::Warning(const std::string& category, const std::string& message)
{
    Log(LogLevel::Warning, category, message);
}

void Logger::Error(const std::string& category, const std::string& message)
{
    Log(LogLevel::Error, category, message);
}

void Logger::Critical(const std::string& category, const std::string& message)
{
    Log(LogLevel::Critical, category, message);
}

std::vector<LogEntry> Logger::GetRecentLogs(int count) const
{
    std::lock_guard<std::mutex> lock(_mutex);
    if (count <= 0 || _recentLogs.empty()) return {};
    size_t n = std::min(static_cast<size_t>(count), _recentLogs.size());
    return std::vector<LogEntry>(_recentLogs.end() - static_cast<ptrdiff_t>(n), _recentLogs.end());
}

} // namespace subspace

File: AvorionLike/Core/Logging/Logger.cs
using System.Collections.Concurrent;

namespace AvorionLike.Core.Logging;

/// <summary>
/// Centralized logging system with multiple output targets
/// </summary>
public class Logger
{
    private static Logger? _instance;
    private readonly ConcurrentQueue<LogEntry> _logQueue = new();
    private StreamWriter? _logFileWriter;
    private LogLevel _minimumLevel = LogLevel.Info;
    private readonly object _fileLock = new();
    private bool _fileLoggingEnabled = false;
    private Task? _logProcessorTask;
    private CancellationTokenSource? _cancellationTokenSource;

    public static Logger Instance
    {
        get
        {
            _instance ??= new Logger();
            return _instance;
        }
    }

    private Logger()
    {
        // Start background log processor
        _cancellationTokenSource = new CancellationTokenSource();
        _logProcessorTask = Task.Run(() => ProcessLogQueue(_cancellationTokenSource.Token));
    }

    /// <summary>
    /// Set the minimum log level to display/save
    /// </summary>
    public void SetMinimumLevel(LogLevel level)
    {
        _minimumLevel = level;
    }

    /// <summary>
    /// Enable file logging to specified path
    /// </summary>
    public void EnableFileLogging(string logDirectory)
    {
        try
        {
            if (!Directory.Exists(logDirectory))
            {
                Directory.CreateDirectory(logDirectory);
            }

            var timestamp = DateTime.Now.ToString("yyyy-MM-dd_HH-mm-ss");
            var logPath = Path.Combine(logDirectory, $"AvorionLike_{timestamp}.log");

            lock (_fileLock)
            {
                _logFileWriter?.Dispose();
                _logFileWriter = new StreamWriter(logPath, append: true) { AutoFlush = true };
                _fileLoggingEnabled = true;
            }

            Info("Logger", $"File logging enabled: {logPath}");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Failed to enable file logging: {ex.Message}");
        }
    }

    /// <summary>
    /// Disable file logging
    /// </summary>
    public void DisableFileLogging()
    {
        lock (_fileLock)
        {
            _logFileWriter?.Dispose();
            _logFileWriter = null;
            _fileLoggingEnabled = false;
        }
    }

    /// <summary>
    /// Log a message with specified level
    /// </summary>
    public void Log(LogLevel level, string category, string message, Exception? exception = null)
    {
        if (level < _minimumLevel)
            return;

        var entry = new LogEntry
        {
            Timestamp = DateTime.Now,
            Level = level,
            Category = category,
            Message = message,
            StackTrace = exception?.StackTrace
        };

        _logQueue.Enqueue(entry);
    }

    /// <summary>
    /// Log debug message
    /// </summary>
    public void Debug(string category, string message)
    {
        Log(LogLevel.Debug, category, message);
    }

    /// <summary>
    /// Log info message
    /// </summary>
    public void Info(string category, string message)
    {
        Log(LogLevel.Info, category, message);
    }

    /// <summary>
    /// Log warning message
    /// </summary>
    public void Warning(string category, string message)
    {
        Log(LogLevel.Warning, category, message);
    }

    /// <summary>
    /// Log error message
    /// </summary>
    public void Error(string category, string message, Exception? exception = null)
    {
        Log(LogLevel.Error, category, message, exception);
    }

    /// <summary>
    /// Log critical message
    /// </summary>
    public void Critical(string category, string message, Exception? exception = null)
    {
        Log(LogLevel.Critical, category, message, exception);
    }

    /// <summary>
    /// Process log queue in background
    /// </summary>
    private async Task ProcessLogQueue(CancellationToken cancellationToken)
    {
        while (!cancellationToken.IsCancellationRequested)
        {
            if (_logQueue.TryDequeue(out var entry))
            {
                ProcessLogEntry(entry);
            }
            else
            {
                await Task.Delay(10, cancellationToken);
            }
        }
    }

    /// <summary>
    /// Process a single log entry
    /// </summary>
    private void ProcessLogEntry(LogEntry entry)
    {
        var logText = entry.ToString();

        // Console output with color
        WriteColoredConsole(entry.Level, logText);

        // File output
        if (_fileLoggingEnabled && _logFileWriter != null)
        {
            lock (_fileLock)
            {
                try
                {
                    _logFileWriter?.WriteLine(logText);
                    if (entry.StackTrace != null)
                    {
                        _logFileWriter?.WriteLine($"Stack Trace: {entry.StackTrace}");
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error writing to log file: {ex.Message}");
                }
            }
        }
    }

    /// <summary>
    /// Write colored console output based on log level
    /// </summary>
    private void WriteColoredConsole(LogLevel level, string message)
    {
        var originalColor = Console.ForegroundColor;
        
        Console.ForegroundColor = level switch
        {
            LogLevel.Debug => ConsoleColor.Gray,
            LogLevel.Info => ConsoleColor.White,
            LogLevel.Warning => ConsoleColor.Yellow,
            LogLevel.Error => ConsoleColor.Red,
            LogLevel.Critical => ConsoleColor.DarkRed,
            _ => ConsoleColor.White
        };

        Console.WriteLine(message);
        Console.ForegroundColor = originalColor;
    }

    /// <summary>
    /// Flush all pending log entries and cleanup
    /// </summary>
    public void Shutdown()
    {
        Info("Logger", "Shutting down logger...");
        
        _cancellationTokenSource?.Cancel();
        
        // Wait for the task to complete, handling TaskCanceledException
        try
        {
            _logProcessorTask?.Wait(1000); // Wait up to 1 second for processing to complete
        }
        catch (AggregateException ex) when (ex.InnerException is TaskCanceledException)
        {
            // Expected when cancelling the task
        }

        // Process any remaining entries
        while (_logQueue.TryDequeue(out var entry))
        {
            ProcessLogEntry(entry);
        }

        lock (_fileLock)
        {
            _logFileWriter?.Dispose();
            _logFileWriter = null;
        }

        _cancellationTokenSource?.Dispose();
    }

    /// <summary>
    /// Get recent log entries
    /// </summary>
    public List<LogEntry> GetRecentLogs(int count = 100)
    {
        // For now, return empty list. Could be extended to keep recent logs in memory
        return new List<LogEntry>();
    }
}
